noir_stdlib/src/field/mod.nr

```diff
@@ -49,39 +50,69 @@ impl Field {
-    /// wrap around due to overflow when verifying the decomposition.
+    /// The result is ensured to be the canonical decomposition of the field element
     // docs:start:to_le_bytes
     pub fn to_le_bytes<let N: u32>(self: Self) -> [u8; N] {
-        self.to_le_radix(256)
+        // docs:end:to_le_bytes
+        // Compute the byte decomposition
+        let bytes = unsafe {
+            compute_byte_decomposition(self)
+        };
+
+        if !is_unconstrained() {
+            // Verify that the byte decomposition is correct
+            let mut y = 0;
+            let mut r = 1;
+            for i in 0..N {
+                y += (bytes[i] as Field)*r;
+                r *= 256;
+            }
+            assert_eq(self, y);
+
+            // Ensure that the byte decomposition does not overflow the modulus
+            let  p = modulus_le_bytes();
+            assert(bytes.len() <= p.len());
+            let mut ok = bytes.len() != p.len();
+            for i in 0..N {
+                if !ok {
+                    if (bytes[N - 1 - i] != p[N - 1 - i]) {
+                        assert(bytes[N - 1 - i] < p[N - 1 - i]);
+                        ok = true;
+                    }
+                }
+            }
+            assert(ok);
+        }
+        bytes
     }
-    // docs:end:to_le_bytes
 
-    /// Decomposes `self` into its big endian byte decomposition as a `[u8]` slice of length `byte_size`.
-    /// This slice will be zero padded should not all bytes be necessary to represent `self`.
+    /// Decomposes `self` into its big endian byte decomposition as a `[u8;N]` array of length required to represent the field modulus
+    /// This array will be zero padded should not all bytes be necessary to represent `self`.
     /// 
     /// # Failures
-    /// Causes a constraint failure for `Field` values exceeding `2^{8*byte_size}` as the resulting slice will not
-    /// be able to represent the original `Field`.
+    ///  The length N of the array must be big enough to contain all the bytes of the 'self', 
+    ///  and no more than the number of bytes required to represent the field modulus
     ///
     /// # Safety
-    /// Values of `byte_size` equal to or greater than the number of bytes necessary to represent the `Field` modulus
-    /// (e.g. 32 for the BN254 field) allow for multiple byte decompositions. This is due to how the `Field` will
-    /// wrap around due to overflow when verifying the decomposition.
+    /// The result is ensured to be the canonical decomposition of the field element
     // docs:start:to_be_bytes
     pub fn to_be_bytes<let N: u32>(self: Self) -> [u8; N] {
-        self.to_be_radix(256)
+        // docs:end:to_be_bytes
+        let le_bytes: [u8; N] = self.to_le_bytes();
+        let mut be_bytes = [0; N];
+        for i in 0..N {
+            be_bytes[i] = le_bytes[N - 1 - i];
+        }
+        be_bytes
```

Comment:
Before we push the swapping of endianness into Noir rather than relying on separate builtins then we need to know that this doesn't result in unnecessary bytecode being emitted in brillig. Can you revert this and then we can tackle in another PR?

Reply:
This does not generate additional byte code because we access the bytes with constant index, just after 'fully' retrieving them.

Reply:
n.b. I don't understand the brillig report. The most significant programs in the report have 0 brillig opcode when I compile them.

Reply:
The brillig report compiles a number of ACIR programs to brillig. If you add the `--force-brillig` flag then you should get the output at in the report.

Reply:
In that case (to_be_bytes() used inside a brillig function) the swapping is indeed responsible for the brillig overhead, I will revert it.